sync15-adapter/src/driver.rs

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// traits etc to drive a "full sync".
+
+use std::result;
+use failure;
+use client::{Sync15StorageClient, Sync15StorageClientInit};
+use state::{GlobalState, SetupStateMachine};
+use sync::{self, Store};
+use key_bundle::KeyBundle;
+
+
+pub trait GlobalStateProvider {
+    fn load(&self) -> result::Result<Option<GlobalState>, failure::Error>;
+
+    fn save(&self, state: Option<&GlobalState>) -> result::Result<(), failure::Error>;
+
+    // Store in memory, but do not persist, a Sync15StorageClientInit
+    fn get_client_init(&self) -> result::Result<Option<Sync15StorageClientInit>, failure::Error>;
+
+    fn set_client_init(&self, init: Option<Sync15StorageClientInit>) -> result::Result<(), failure::Error>;
+}
+
+pub fn sync(
+    store: &Store,
```

Comment:
I don't love this approach, but if it's the one we're going to take, we should take a slice of stores, probably.

This function also should return `Result<()>`.

Also, a nit: this should be `&dyn Store` and `&dyn GlobalStateProvider`. Using trait objects without indicating they're trait objects will be removed in the 2018 edition, and IMO it makes what's happening less clear.